sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+    objectives = _get_objective_nodes(state_sequences)
+
+    for victim in total_victims:
+        print('!!! Rendering AGs for Victim', victim)
+        for objective in objectives:
+            # Get the variants of current objective and attempts per team
+            team_attack_attempts, observed_obj = _get_attack_attempts(state_sequences, victim, objective, in_main_model)
+
+            # If no team has obtained this objective or has targeted this victim, don't generate its AG
+            if sum([len(attempt) for attempt in team_attack_attempts.values()]) == 0:
                 continue
 
-            AGname = attack.replace('|', '').replace('_','').replace('-','').replace('(','').replace(')', '')
-            lines = []
-            lines.append((0,'digraph '+ AGname + ' {'))
-            lines.append((0,'rankdir="BT"; \n graph [ nodesep="0.1", ranksep="0.02"] \n node [ fontname=Arial, fontsize=24,penwidth=3]; \n edge [ fontname=Arial, fontsize=20,penwidth=5 ];'))
-            root_node = translate(attack, root=int_victim)
-            lines.append((0, '"'+root_node+'" [shape=doubleoctagon, style=filled, fillcolor=salmon];'))
-            lines.append((0, '{ rank = max; "'+root_node+'"}'))
-            
-            for obj in list(observed_obj): # for each variant of objective, add a link to the root node, and determine if it's sink
-                lines.append((0,'"'+translate(obj)+'" -> "'+root_node+'"'))
-                
-                sinkflag = False
-                for sink in sev_sinks:    
-                    if obj.split("|")[-1] == sink:
-                        sinkflag = True
-                        break
-                if sinkflag:
-                    lines.append((0,'"'+translate(obj)+'" [style="filled,dotted", fillcolor= salmon]'))
+            print('\t!!!! Objective', objective)
+            ag_name = objective.replace('|', '').replace('_', '').replace('-', '').replace('(', '').replace(')', '')
+            lines = ['digraph ' + ag_name + ' {',
+                     'rankdir="BT"; \n graph [ nodesep="0.1", ranksep="0.02"] \n node [ fontname=Arial, ' +
+                     'fontsize=24, penwidth=3]; \n edge [ fontname=Arial, fontsize=20,penwidth=5 ];']
+            root_node = _translate(objective, root=victim)
+            lines.append('"' + root_node + '" [shape=doubleoctagon, style=filled, fillcolor=salmon];')
+            lines.append('{ rank = max; "' + root_node + '"}')
+
+            # For each variant of objective, add a link to the root node, and determine if it's sink
+            for obj_variant in list(observed_obj):
+                lines.append('"' + _translate(obj_variant) + '" -> "' + root_node + '"')
+                if _is_severe_sink(obj_variant, sev_sinks):
+                    lines.append('"' + _translate(obj_variant) + '" [style="filled,dotted", fillcolor= salmon]')
                 else:
-                    lines.append((0,'"'+translate(obj)+'" [style=filled, fillcolor= salmon]'))
```

Comment:
These 0's in the `lines` were not used anywhere and I haven't thought of any ways they could be used

Reply:
This is explicitly for the generation of the .dot file. 0 makes sure these lines are at the top when we sort the list.

Reply:
Oh, but the list with lines was not sorted anywhere. In this case, do you me still add these 0's back **and** sort the list? Because currently AGs are generated fine

Reply:
Revert this change. We keep the tuple format anyway.

Reply:
Done